In [ ]:
# Product Category Prediction

import numpy as np
import pandas as pd
import joblib

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

In [ ]:
# Load dataset

import pandas as pd


df =pd.read_csv(r"C:\Users\HP\OneDrive\Desktop\svm\project\product_sales_dataset_final.csv")


In [ ]:
# 1. CHECK BALANCE

print(df['Category'].value_counts())

Category
Clothing & Apparel    62298
Electronics           51230
Home & Furniture      50564
Accessories           35908
Name: count, dtype: int64


In [ ]:
# 2. BALANCE DATASET

df_ele   = df[df['Category'] == 'Electronics'].sample(20000, random_state=42)
df_cloth = df[df['Category'] == 'Clothing & Apparel'].sample(20000, random_state=42)
df_home  = df[df['Category'] == 'Home & Furniture'].sample(20000, random_state=42)
df_acc   = df[df['Category'] == 'Accessories'].sample(20000, random_state=42)

df_new = pd.concat([df_ele, df_cloth, df_home, df_acc])
df_new = df_new.sample(frac=1, random_state=42)

print(df_new['Category'].value_counts())

Category
Home & Furniture      20000
Accessories           20000
Electronics           20000
Clothing & Apparel    20000
Name: count, dtype: int64


In [ ]:
df_new.columns = df_new.columns.str.strip().str.lower()

In [ ]:
df_new

order_id order_date      customer_name              city  \
25444      25445   11-09-23     Michael Wilson          Chandler   
16100      16101   05-14-24       Max Franklin           Lincoln   
181814    181815   06-29-23    Melanie Simpson            Dallas   
82760      82761   11-13-23       Jeffrey Hill              Mesa   
27394      27395   06-30-23        Luis Gordon  Colorado Springs   
...          ...        ...                ...               ...   
111967    111968   01-22-23     Tammy Mcdowell            Austin   
123396    123397   05-04-24  Julie Christensen          Chandler   
89201      89202   05-29-24       Laura Turner       Albuquerque   
103621    103622   06-29-23       Ashley Lewis        Pittsburgh   
173633    173634   12-25-23      Gregory Flynn       San Antonio   

               state  region        country          category sub_category  \
25444        Arizona    West  United States  Home & Furniture    Furniture   
16100       Nebraska  Centre  United States  Home & Furniture   Home Decor   
181814         Texas   South  United States       Accessories         Bags   
82760        Arizona    West  United States       Accessories         Bags   
27394       Colorado    West  United States  Home & Furniture  Kitchenware   
...              ...     ...            ...               ...          ...   
111967         Texas   South  United States       Electronics    Wearables   
123396       Arizona    West  United States  Home & Furniture      Bedding   
89201     New Mexico    West  United States       Accessories         Bags   
103621  Pennsylvania    East  United States       Electronics      Laptops   
173633         Texas   South  United States       Electronics      Tablets   

                 product_name  quantity  unit_price  revenue   profit  
25444            Office Chair         1      368.11   368.11    79.30  
16100           Throw Pillows         3      312.29   936.87   211.26  
181814                 Wallet         2       70.10   140.20    26.20  
82760                  Wallet         1      200.72   200.72    38.85  
27394            Cookware Set         5      853.00  4265.00   672.23  
...                       ...       ...         ...      ...      ...  
111967            Apple Watch         6      934.78  5608.68   614.91  
123396  Tempur-Pedic Mattress         4      880.13  3520.52  1042.17  
89201                Tote Bag         2       95.76   191.52    55.88  
103621            Dell XPS 13         1      694.87   694.87    69.25  
173633               iPad Pro         4      698.31  2793.24   479.37  

[80000 rows x 14 columns]

In [ ]:
df_new['unit_price'] = df_new['revenue'] / df_new['quantity']

In [ ]:
margin_map = {
    "Electronics": 0.15,          # 15%
    "Clothing & Apparel": 0.60,   # 60%
    "Home & Furniture": 0.35,     # 35%
    "Accessories": 0.50           # 50%
}


In [ ]:
df_new["purchase_price"] = df_new.apply(
    lambda row: row["unit_price"] * (1 - margin_map[row["category"]]),
    axis=1
)

In [ ]:
df_new["revenue"] = df_new["quantity"] * df_new["unit_price"]
df_new["profit"] = df_new["revenue"] - (df_new["quantity"] * df_new["purchase_price"])

In [ ]:
df_new.head()

order_id order_date    customer_name              city     state  \
25444      25445   11-09-23   Michael Wilson          Chandler   Arizona   
16100      16101   05-14-24     Max Franklin           Lincoln  Nebraska   
181814    181815   06-29-23  Melanie Simpson            Dallas     Texas   
82760      82761   11-13-23     Jeffrey Hill              Mesa   Arizona   
27394      27395   06-30-23      Luis Gordon  Colorado Springs  Colorado   

        region        country          category sub_category   product_name  \
25444     West  United States  Home & Furniture    Furniture   Office Chair   
16100   Centre  United States  Home & Furniture   Home Decor  Throw Pillows   
181814   South  United States       Accessories         Bags         Wallet   
82760     West  United States       Accessories         Bags         Wallet   
27394     West  United States  Home & Furniture  Kitchenware   Cookware Set   

        quantity  unit_price  revenue     profit  purchase_price  
25444          1      368.11   368.11   128.8385        239.2715  
16100          3      312.29   936.87   327.9045        202.9885  
181814         2       70.10   140.20    70.1000         35.0500  
82760          1      200.72   200.72   100.3600        100.3600  
27394          5      853.00  4265.00  1492.7500        554.4500

In [ ]:
# 3. FEATURES & TARGET

X = df_new[['quantity', 'unit_price', 'purchase_price' ,'revenue', 'profit']]
y = df_new['category']

In [ ]:
# 4. ENCODING

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y_enc = le.fit_transform(y)

In [ ]:
# 5. SCALING (MANDATORY FOR SVM)

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [ ]:
# 6. TRAIN TEST SPLIT

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y_enc, test_size=0.2, random_state=42
)


In [ ]:
# Feature Selection using RFECV
# RFECV (Recursive Feature Elimination with Cross-Validation) is a powerful technique for selecting the most relevant features for a model. It works by recursively removing features and evaluating the model's performance using cross-validation to determine the optimal number of features.

from sklearn.feature_selection import RFECV
from sklearn.model_selection import StratifiedKFold
from sklearn.svm import SVC
from sklearn.svm import LinearSVC

# enable probability so scorer can use predict_proba (optional but safe)
svc = SVC(
    kernel='rbf',
    C=10,
    gamma=0.1,
    class_weight='balanced',
    probability=True
)

rfecv = RFECV(
    estimator=svc,
    step=1,
    cv=StratifiedKFold(5),
    scoring='roc_auc_ovr',   # use multiclass-compatible scorer ('ovr' or 'ovo')
    n_jobs=-1
)


# RFECV needs an estimator that exposes coef_ or feature_importances_
linear_svc = LinearSVC(C=1.0, class_weight='balanced', max_iter=10000)
rfecv.estimator = linear_svc


# LinearSVC has no predict_proba — use a scorer that doesn't require probabilities
rfecv.scoring = 'f1_weighted'   # alternatives: 'accuracy', 'balanced_accuracy'
rfecv.fit(X_train, y_train)

print("Optimal number of features:", rfecv.n_features_)


Optimal number of features: 2


In [ ]:
X_train_selected = rfecv.transform(X_train)
X_test_selected = rfecv.transform(X_test)

print("Before:", X_train.shape)
print("After :", X_train_selected.shape)


Before: (64000, 5)
After : (64000, 2)


In [ ]:
# 7. MODEL TRAINING

from sklearn.svm import SVC

model = SVC(
    kernel='rbf',
    C=10,
    gamma=0.1,
    class_weight='balanced',
    probability=True
)

model.fit(X_train_selected, y_train)

SVC(C=10, class_weight='balanced', gamma=0.1, probability=True)

In [ ]:
# 8. EVALUATION

from sklearn.metrics import classification_report, accuracy_score

y_pred = model.predict(X_test_selected)
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.99975
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      4008
           1       1.00      1.00      1.00      3934
           2       1.00      1.00      1.00      3982
           3       1.00      1.00      1.00      4076

    accuracy                           1.00     16000
   macro avg       1.00      1.00      1.00     16000
weighted avg       1.00      1.00      1.00     16000



In [ ]:
# 9. SAVE MODEL

# joblib already imported at the top of the notebook
joblib.dump(rfecv, "rfecv_selector.pkl")
joblib.dump(model, "svm_model.pkl")
joblib.dump(scaler, "scaler.pkl")
# use the fitted LabelEncoder instance `le`
joblib.dump(le, "label_encoder.pkl")

print("Model, scaler, encoder saved successfully")

Model, scaler, encoder saved successfully


check krne ke liye  : 

* **Quantity = 2**
* **Unit_price = 203**
* **purchase_price = 110**

* **output : accessories**

# --------------------------------------------------------

* **Quantity = 4**
* **Unit_price = 377**
* **purchase_price = 322**

* **output : different category**